From 663c4452f37dd6b10175c165a964e659e81506e8 Mon Sep 17 00:00:00 2001
From: Rui Chen <rui@chenrui.dev>
Date: Thu, 26 Dec 2024 02:13:57 -0500
Subject: [PATCH] fix: remove unused to_string_view overload in fmt >= 11.1
 (#3314)

fix: update to_string_view function for fmt 11.1 (#3301)

Signed-off-by: Rui Chen <rui@chenrui.dev>

fix: Compatibility with external fmtlib 11.1.1 (#3312)

Include fmtlib's `xchar` header to include `fmt::basic_format_string`.
Otherwise, compilation with an external fmtlib 11.1.1 fails with

```
In file included from include/spdlog/spdlog.h:12:
include/spdlog/common.h:369:49: error: no template named 'basic_format_string' in namespace 'fmt'; did you mean 'std::basic_format_string'?
  369 | inline fmt::basic_string_view<T> to_string_view(fmt::basic_format_string<T, Args...> fmt) {
      |                                                 ^~~~~
```

Signed-off-by: Christian Blichmann <cblichmann@google.com>
---
 include/spdlog/common.h  | 7 +------
 include/spdlog/fmt/fmt.h | 1 +
 2 files changed, 2 insertions(+), 6 deletions(-)

diff --git a/include/spdlog/common.h b/include/spdlog/common.h
index aca483c2..71ffd240 100644
--- a/include/spdlog/common.h
+++ b/include/spdlog/common.h
@@ -364,12 +364,7 @@ SPDLOG_CONSTEXPR_FUNC spdlog::wstring_view_t to_string_view(spdlog::wstring_view
 }
 #endif
 
-#ifndef SPDLOG_USE_STD_FORMAT
-template <typename T, typename... Args>
-inline fmt::basic_string_view<T> to_string_view(fmt::basic_format_string<T, Args...> fmt) {
-    return fmt;
-}
-#elif __cpp_lib_format >= 202207L
+#if defined(SPDLOG_USE_STD_FORMAT) &&  __cpp_lib_format >= 202207L
 template <typename T, typename... Args>
 SPDLOG_CONSTEXPR_FUNC std::basic_string_view<T> to_string_view(
     std::basic_format_string<T, Args...> fmt) SPDLOG_NOEXCEPT {
diff --git a/include/spdlog/fmt/fmt.h b/include/spdlog/fmt/fmt.h
index 7fa6b093..2f09c15e 100644
--- a/include/spdlog/fmt/fmt.h
+++ b/include/spdlog/fmt/fmt.h
@@ -27,4 +27,5 @@
 #else  // SPDLOG_FMT_EXTERNAL is defined - use external fmtlib
     #include <fmt/core.h>
     #include <fmt/format.h>
+    #include <fmt/xchar.h>
 #endif
-- 
2.47.1.windows.1

